// Copyright (c) Huawei Technologies Co., Ltd. 2021. All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_package = "ohos.devtools.datasources.transport.grpc.service";
option optimize_for = LITE_RUNTIME;

import "memory_plugin_common.proto";

// Data format of memory collect plug-in.
// Various memory data count, including system level and application level.
// Obtained from the proc file system or system service.
message SysMeminfo {
    SysMeminfoType key = 1;
    uint64 value = 2;
};
message SysVMeminfo {
    SysVMeminfoType key = 1;
    uint64 value = 2;
};

message SmapsInfo {
    string start_addr = 1;
    string end_addr = 2;
    string permission = 3;
    string path = 4;
    uint64 size = 5;
    uint64 rss = 6;
    uint64 pss = 7;
    double reside = 8;
    uint64 dirty = 9;
    uint64 swapper = 10;
    uint64 private_clean = 11;
    uint64 private_dirty = 12;
    uint64 shared_clean = 13;
    uint64 shared_dirty = 14;
    uint64 swap = 15;
    uint64 swap_pss = 16;
    string category = 17;
};

message AppSummary {
    uint64 java_heap = 1;
    uint64 native_heap = 2;
    uint64 code = 3;
    uint64 stack = 4;
    uint64 graphics = 5;
    uint64 private_other = 6;
    uint64 system = 7;
};

message ProcessMemoryInfo {
    int32 pid = 1;
    string name = 2;
    // data from /proc/$pid/stat
    uint64 vm_size_kb = 3;
    uint64 vm_rss_kb = 4;
    uint64 rss_anon_kb = 5;
    uint64 rss_file_kb = 6;
    uint64 rss_shmem_kb = 7;
    uint64 vm_swap_kb = 8;
    uint64 vm_locked_kb = 9;
    uint64 vm_hwm_kb = 10;
    int64 oom_score_adj = 11;
    // data from /proc/$pid/smaps
    repeated SmapsInfo smapinfo = 12;
    AppSummary memsummary = 13;
}

message MemoryData {
    repeated ProcessMemoryInfo processesinfo = 1;
    repeated SysMeminfo meminfo = 2;
    repeated SysVMeminfo vmeminfo = 3;
    uint64 zram = 4;
}
